lib/SecretsCrypto/secrets.c

```diff
@@ -0,0 +1,1000 @@
+		if (error->code == G_IO_ERROR_CLOSED) {
+			struct Sailfish_Secrets_Result *result =
+			    Sailfish_Secrets_Result_new
+			    (Sailfish_Secrets_Result_Succeeded, 0, "");
+			struct Sailfish_Secrets_Callback_Data *cbd =
+			    (struct Sailfish_Secrets_Callback_Data *)user_data;
+			g_object_unref(daemon_connection.secretsProxy);
+			daemon_connection.secretsProxy = NULL;
+			g_object_unref(daemon_connection.p2pBus);
+			daemon_connection.p2pBus = NULL;
+			daemon_connection.busy = 0;
+			cbd->result_callback(cbd->callback_context, result);
+			Sailfish_Secrets_Result_unref(result);
+			Sailfish_Secrets_Callback_Data_unref(cbd);
+		} else {
+			struct Sailfish_Secrets_Result *result =
+			    Sailfish_Secrets_Result_new
+			    (Sailfish_Secrets_Result_Failed, 2,
+			     "Unable to disconnect from the secrets daemon");
+			struct Sailfish_Secrets_Callback_Data *cbd =
+			    (struct Sailfish_Secrets_Callback_Data *)user_data;
+			daemon_connection.busy = 0;
+			cbd->result_callback(cbd->callback_context, result);
+			Sailfish_Secrets_Result_unref(result);
+			Sailfish_Secrets_Callback_Data_unref(cbd);
+		}
+		g_error_free(error);
+	}
+}
+
+/******************************* Internal DBus Marshalling ******************/
+
+GVariant *Sailfish_Secrets_variantFromSecretIdentifier(
+		struct Sailfish_Secrets_Secret_Identifier *ident)
+{
+	return g_variant_new("(ss)", ident->name, ident->collectionName);
+}
+
+GVariant *Sailfish_Secrets_variantFromSecret(
+		struct Sailfish_Secrets_Secret *secret)
+{
+	GVariant *result = NULL;
+	GVariantBuilder *dataBuilder = NULL;
+	GVariantBuilder *filterDataBuilder = NULL;
+	struct Sailfish_Secrets_Secret_FilterDatum *currFilter = NULL;
+	size_t i = 0;
+
+	if (secret == NULL)
+		return NULL;
+
+	dataBuilder = g_variant_builder_new(G_VARIANT_TYPE_BYTESTRING);
+	for (i = 0; i < secret->dataSize; i++)
+		g_variant_builder_add(dataBuilder, "y", secret->data[i]);
+
+	filterDataBuilder = g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+	currFilter = secret->filterData;
+	while (currFilter) {
+		g_variant_builder_add(filterDataBuilder, "{sv}",
+				      currFilter->field,
+				      g_variant_new("s", currFilter->value));
+		currFilter = currFilter->next;
+	}
+
+	result = g_variant_new(
+			    "(@(ss)@ay@a{sv})",
+			    Sailfish_Secrets_variantFromSecretIdentifier(secret->identifier),
+			    g_variant_builder_end(dataBuilder),
+			    g_variant_builder_end(filterDataBuilder));
+
+	g_variant_builder_unref(dataBuilder);
+	g_variant_builder_unref(filterDataBuilder);
+
+	return result;
+}
+
+struct Sailfish_Secrets_Secret *
+Sailfish_Secrets_secretFromVariant(GVariant *variant)
+{
+	struct Sailfish_Secrets_Secret *secret = NULL;
+	GVariant *identVariant = NULL;
+	char *secretName = NULL;
+	char *collectionName = NULL;
+	GVariantIter *dataIter;
+	unsigned char datum = 0;
+	unsigned char *data =
+	    (unsigned char *)malloc(2048 * sizeof(unsigned char));
```

Comment:
What does 2048 means? What if we'll need to increase this value? Change it everywhere where we meet it?
`#define NEEDED_BYTES 2048` ???

Reply:
Urgh, I should update this to use GByteArray and the from_fixed_array accessor, as I did in crypto.c.  This entire block is suboptimal and error prone.  I will update.